packages/nimbus/src/docs/architecture-decisions/adr-0004-component-api-patterns.mdx

```diff
@@ -0,0 +1,378 @@
+
+**The cost-benefit:** Slots are powerful but add cognitive load. Only pay this cost when consumers genuinely need the flexibility.
+
+**If YES:** Use **Compound Components + Slots**  
+**If NO:** Use **Standard Compound Components**
+
+## Real Examples
+
+### Button → Single Component ✅
+
+**Decision path:** One thing? ✅ → Different spots? ❌
+
+**Real codebase usage:**
+
+```tsx
+<Button variant="primary">Save</Button>
+<Button variant="subtle">Cancel</Button>
+```
+
+**Why Single Component:** Only one content area (text or icon), simple internal
+implementation.
+
+### DatePicker → Single Component ✅
+
+**Decision path:** One thing? ✅ → Different spots? ❌
+
+**Real codebase usage:**
+
+```tsx
+<DatePicker
+  aria-label="Select a date"
+  value={selectedDate}
+  onChange={setSelectedDate}
+/>
+```
+
+**Why Single Component:** One conceptual input (date selection). Internal
+complexity (calendar, dropdowns, validation) is hidden from consumers.
+
+### Badge → Single Component ✅
+
+**Decision path:** One thing? ✅ → Different spots? ❌
+
+**Real codebase usage:**
+
+```tsx
+<Badge size="md" colorPalette="blue">
+  Demo Badge
+</Badge>
+```
+
+**Why Single Component:** Only one content area (text), simple internal
+implementation.
+
+### Card → Single Component + Slots ⚠️
+
+**Decision path:** One thing? ✅ → Different spots? ✅
+
+```tsx
+<!-- Current (compound pattern) -->
+<Card.Root cardPadding="md" borderStyle="outlined">
+  <Card.Header><b>Card Title</b></Card.Header>
+  <Card.Content><p>Card content here</p></Card.Content>
+</Card.Root>
+
+<!-- Should be (single component + slots) -->
+<Card cardPadding="md" borderStyle="outlined">
+  <Text slot="header">Card Title</Text>
+  <Stack slot="content">
+    <Text>Card content here</Text>
+    <Button>Action</Button>
+  </Stack>
+</Card>
+```
+
+### Select → Compound + Slots ✅
+
+**Decision path:** One thing? ❌ → Different jobs? ✅ → Flexible content? ✅
+
+```tsx
+<Select.Root>
+  <Select.Options>
+    <Select.Option>Simple</Select.Option>
+    <Select.Option>
+      <Text slot="label">Rich</Text>
+      <Text slot="description">With description</Text>
+    </Select.Option>
+  </Select.Options>
+</Select.Root>
+```
+
+### Accordion → Compound + Slots ⚠️
+
+**Decision path:** One thing? ❌ → Different jobs? ✅ → Flexible content? ✅
+
+```tsx
+<!-- Current (mixed compound + internal slots) -->
+<Accordion.Root>
+  <Accordion.Item>
+    <Accordion.Header>Simple text</Accordion.Header>
+    <Accordion.Content>Content here</Accordion.Content>
+  </Accordion.Item>
+</Accordion.Root>
+
+<!-- Should be (true compound + slots pattern) -->
+<Accordion.Root>
+  <Accordion.Item>
+    <Text slot="header">Simple text</Text>
+    <Text slot="content">Content here</Text>
+  </Accordion.Item>
+  <Accordion.Item>
+    <Stack slot="header">
+      <Text>Complex Header</Text>
+      <Button>Edit</Button>
+    </Stack>
+    <Text slot="content">More content</Text>
```

Comment:
We have to be careful here - to use this as a `react-aria`-style `slot`, the component needs to be consuming a context. 

This works for the `Text` components, because our `Text` component [consumes](https://github.com/commercetools/nimbus/blob/df0d648b439697177a2f8861c81b2da17974f434/packages/nimbus/src/components/text/text.tsx#L20) `react-aria`'s `TextContext`.

The `Stack` on the other hand does not consume a context, so targeting it with a slot would not do much as things stand currently. We would have to implement a custom `SlotContext` to achieve this particular example.